### PR TITLE
major uplift, drop support for python < 3.7 and migrate to using mesh-sandbox

### PR DESCRIPTION
**Pull request by matt-mercer**
Aligning it with other MESH repos, 
- use pyproject.toml
- use poetry, asdf, pytest, flake8, black, isort, coverage
- Major change: drop support for End of Life python versions ( pre python 3.7 )
- Major change: remove MockMeshServer from wheel distribution ( with an intent to remove it completely in the near future)
- Add type annotations, where possible

### CODE REVIEW

**Comment by jamespic** [comment]
The changes are definitely far reaching, but look reasonable as far as I can see (and hopefully someone closer to the ongoing work has reviewed it in any case).

The only thing that I think is worth noting is that some of these changes are breaking - most obviously, dropping Python 2 support, which I know some of the projects that depend on Mesh Client are still using. It looks like we're doing the right thing with Python version specs (so if a new version is published then it shouldn't get installed on Python 2), but it's probably worth reaching out to some of the teams who are using this, and it's probably worth a major version bump to keep things semver.

________________________________
From: Matthew Mercer ***@***.***>
Sent: Thursday, June 29, 2023 11:16:03 pm
To: NHSDigital/mesh-client ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [NHSDigital/mesh-client] mesh-1478: align with other projects use pyproject and tools versions (PR #7)

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/NHSDigital/mesh-client/pull/7

Commit Summary

  *   2717e59<https://github.com/NHSDigital/mesh-client/pull/7/commits/2717e59298cb932b4f6539da06eaf007b3a9f67f> mesh-1478: align with other projects use pyproject and tools versions

File Changes

(35 files<https://github.com/NHSDigital/mesh-client/pull/7/files>)

  *   A .gitallowed<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-58f84f1ca7c1d73093ee12b7c3c67032877b926e66d3f909ed7eb6b0537886ec> (5)
  *   A .gitdisallowed<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-23d974fb05ef1e393ee4416116620bcc5b5fd3eb944a8b21aef957a4c064b74b> (21)
  *   A .github/CODEOWNERS<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-3d36a1bf06148bc6ba1ce2ed3d19de32ea708d955fed212c0d27c536f0bd4da7> (1)
  *   A .github/dependabot.yml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-dd4fbda47e51f1e35defb9275a9cd9c212ecde0b870cba89ddaaae65c5f3cd28> (11)
  *   A .github/workflows/merge-develop.yml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-daeb1ac4d246d09ba839b2dccaa2ea2d2d76340db6fe91c9b599f6f256941740> (159)
  *   A .github/workflows/pull-request.yml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-b71166ed0f585913318ed46933ff9b12901e211de3ac88c40de03f0a944c0ae0> (313)
  *   D .github/workflows/python-package.yml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-ee49282f461b4c8ad179f79dd5bcdf93124561074c64a771366caf93e99b9320> (26)
  *   M .tool-versions<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-751af1a340658c7b8176fe32d7db9fadbe15d1d075daba1919a91df04155bc70> (6)
  *   A CONTRIBUTING.md<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055> (79)
  *   M LICENSE<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-c693279643b8cd5d248172d9c22cb7cf4ed163a3c98c8a3f69c2717edd3eacb7> (22)
  *   M Makefile<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52> (111)
  *   M README.md<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5> (6)
  *   D dev-requirements.txt<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-43470c4f399e798afc682f0bf6c690eef5f47ee25445904a066e4c3314f3a787> (8)
  *   A git-secrets<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-dfeb3e3e275364adae7b60e28955422400916e5a86d339674ac7113e8b4a9081> (357)
  *   M mesh_client/__init__.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-3a86b0b77f842fb4ed2e780ab3b68ebbbe87c314d3f8aefb1f4351ce178b289e> (237)
  *   M mesh_client/io_helpers.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-e3002c1f7efc54e5debe4d3090d76d35c5807e1c69c6805bf56689216aaf1c98> (33)
  *   M mesh_client/key_helper.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-818bd061993f71f574cebecf4dece39a6794814d6e52c21457a91f573034192b> (15)
  *   M mesh_client/mesh_auth.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-95693ae29d07941b9b1b32951a530da435cb82725712ddbb5f5559fe926302e7> (22)
  *   M mesh_client/mock_server.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-ee8a4ed2a509ae65b15fbe2ddb30b09c5fc348aab0497f42fc87a3a7b9fb5a5a> (204)
  *   A mesh_client/py.typed<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-bc821f8e7b461d19e1d9e99a309168f73341ad9e74ef18b5e4ec4afaba5244a8> (0)
  *   A poetry.lock<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-f53a023eedfa3fbf2925ec7dc76eecdc954ea94b7e47065393dbad519613dc89> (1444)
  *   A poetry.toml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-7f506357578ba48a673f1dde44cad7d64eb564eeb286716496e7776cdbb2539c> (4)
  *   M pyproject.toml<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711> (147)
  *   A scripts/check-secrets.sh<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-019c27b7e1851c31b7dff95fc6645cb2aabe7d3642de7128aca156872fb6721c> (17)
  *   A scripts/hooks/commit-msg.sh<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-493c88d6a51fd76acb5807631b60ba1847098e702cf2c412ac2637ac34678db4> (17)
  *   A scripts/hooks/pre-commit.sh<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-3901cbf6c2e7a4fb6dc47254038e7b20110405697ae939514de453c41983f809> (31)
  *   A scripts/sonar_tests.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-2a2b5eda85dab51528f0473a32abb63e33cc91d4d46c2f17eb507cf1fbb202ee> (161)
  *   A scripts/wait-for-container.sh<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-e9416642bd3ca1fd5888f41087cd35742e503a7e4962a3bce105b070c887e55b> (14)
  *   M setup.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7> (65)
  *   A tests/__init__.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-001e61d97d9dee27d0c5aa1f23ba6c1972cd8a2e8320bac839656b0ab935b84d> (0)
  *   R tests/io_helpers_test.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-cbad110bb72c1e2182789598b777b98f0bd443b462364576a625565abf75a086> (101)
  *   R tests/mesh_client_test.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-1feeb1994f8bd74f5e249ce93000137aebfe9b2bae7c90b3ca4d9bb55db6375b> (164)
  *   R tests/mesh_endpoint_connectivity_test.py<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-bc518e2839d49b7b2ec64a5d679b8cf12501ea19f50d385d03f98a2b0283bc66> (10)
  *   A tests/py.typed<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-bc9826e6a6404b71b61c86892c2a52a45c7e74562acab4b0c4153af70cbbdfdb> (0)
  *   D tox.ini<https://github.com/NHSDigital/mesh-client/pull/7/files#diff-ef2cef9f88b4fe09ca3082140e67f5ad34fb65fb6e228f119d3812261ae51449> (7)

Patch Links:

  *   https://github.com/NHSDigital/mesh-client/pull/7.patch
  *   https://github.com/NHSDigital/mesh-client/pull/7.diff

**Comment by matt-mercer** [comment]
> The changes are definitely far reaching, but look reasonable as far as I can see (and hopefully someone closer to the ongoing work has reviewed it in any case). The only thing that I think is worth noting is that some of these changes are breaking - most obviously, dropping Python 2 support, which I know some of the projects that depend on Mesh Client are still using. It looks like we're doing the right thing with Python version specs (so if a new version is published then it shouldn't get installed on Python 2), but it's probably worth reaching out to some of the teams who are using this, and it's probably worth a major version bump to keep things semver.


Absolutely .. this is definitely a MAJOR release .. PYTHON 2.7 was EOL over 9 years ago .. even python 3.7 is EOL .. so I think this is being 'cautious' in maintaining 3.7 support .. (I would have really liked to use some 3.8+ type hints, but I've had to pull back from that for now)

File: mesh_client/__init__.py [hunk 1]
```diff
@@ -337,59 +409,66 @@ def send_message(
         compression for sending is enabled as a constructor option.
```
Comment (johnwilliams57-nhs) [inline mesh_client/__init__.py:413]: i think this doc string needs updating, removing the partner_id, and smtp headers

File: mesh_client/__init__.py [hunk 2]
```diff
@@ -472,7 +553,29 @@ def __exit__(self, type_, value, tb):
...
+    local_id: Optional[str] = None
+    partner_id: Optional[str] = None
+    chunk_range: Optional[str] = None
+
+    subject: Optional[str] = None
+    encrypted: Optional[Union[str, bool]] = None
+    compressed: Optional[Union[str, bool]] = None
+
+
+class _BaseMessage:
```
Comment (johnwilliams57-nhs) [inline mesh_client/__init__.py:586]: again the doc string looks like it needs updating here. missing checksum i think

**Review by johnwilliams57-nhs — COMMENTED** [review]
just adding a couple of comments about doc strings, which i think need tweaking

File: mesh_client/__init__.py (hunk 1 above)
Comment (matt-mercer) [reply mesh_client/__init__.py:413]: partner_id isn't going .. it's process_id that's been removed ..  (but yes .. that was left behind in docstrings!)

File: mesh_client/__init__.py (hunk 2 above)
Comment (matt-mercer) [reply mesh_client/__init__.py:586]: thanks

**Review by davidhallam4-nhs — APPROVED** [review]

**Pull request merged**